esx_service/vmdk_ops.py

```diff
@@ -862,83 +874,114 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts, vc_u
                   vm_datastore, vm_datastore_url)
 
     error_info, tenant_uuid, tenant_name = auth.get_tenant(vm_uuid)
+    force_detach = False
     if error_info:
-        # For docker volume ls, docker prints a list of cached volume names in case
-        # of error(in this case, orphan VM). See Issue #990
-        # Explicity providing empty list of volumes to avoid misleading output.
+        # For "docker volume ls", in case of error fron the plugin Docker prints a list of cached volume names,
+        # which is misleading. To avoid this, we replace error with an empty list. See Issue #990 for details.
         if (cmd == "list") and (not tenant_uuid):
             return []
+        # We need special handling for failure to find tenants in "detach".
+        # get_tenant() will  fail if the VM was removed from the vmgroup or the vmgroup was deleted
+        # by the admin while the disk was attached. See issue #1441 for one of the cases.
+        elif (cmd =="detach"):
+            force_detach = True
         else:
             return err(error_info)
 
-    # default_datastore must be set for tenant
-    error_info, default_datastore_url = auth_api.get_default_datastore_url(tenant_name)
-    if error_info:
-        return err(error_info.msg)
-    elif not default_datastore_url:
-        err_msg = error_code_to_message[ErrorCode.DS_DEFAULT_NOT_SET].format(tenant_name)
-        logging.warning(err_msg)
-        return err(err_msg)
+    if force_detach:
+        # Special (ugly) patch for detaching from VMs where we can't find tenant
+        # (e.g. tenant definition was removed)
+        # The patch is ugly since the affected code is a bit convoluted and can benefit
+        # from refactoring.
+        # The patch does the following: circumvents all the code for authentication and
+        # vmdk path calculation, and instead find good match in the list of devices actually attached.
+
+        logging.warning("executeRequest: FORCE_DETACH vm_uuid=%s, vm_name=%s, full_volume_name=%s",
+                        vm_uuid, vm_name, full_vol_name)
+        # For detach, we get full volume name from docker so it should always be valid.
+        try:
+            vol_name, datastore = parse_vol_name(full_vol_name)
+            logging.info("vol_name=%s, datastore=%s", vol_name, datastore)
+        except ValidationError as ex:
+            return err(str(ex))
+        # we use tenant name to form a unique lock name, so let's fake it
+        tenant_name = "__NoSuchTenant__"
 
-    # default_datastore could be a real datastore name or a hard coded  one "_VM_DS"
-    default_datastore = get_datastore_name(default_datastore_url)
+        # Since we do not know the tenant and thus cannot construct the /vmfs/volumes/<datastore>/dockvols/<tenant>
+        # let's just look in the attached device for the best match.
+        vmdk_path = vmdk_utils.get_attached_volume_path(vm, vol_name, datastore)
```

Comment:
+1 @govint 

> msterin> impossible. datastore is provided by Docker on detach and is always populated. But I agree, it's better to check for this

Now I recall. We convert all short names to fully qualified ones on plugin side inside docker host. Hence it will always be full names.
So for detach due to an unmount, we will always have full name

But for detach as part of create volume work flow ( wiz create, attach, format detach), can we get volume name short? Either way deleting a vmgroup just in the middle of volume creation looks very very rare :))